src/typing/env_js.ml

```diff
@@ -75,33 +80,58 @@ let find_env ?(for_type=false) cx x reason =
   let rec loop = function
     | [] -> find_global cx x reason
     | block::blocks ->
-        (match SMap.get x !block with
+        let entries = block.entries in
+        (match SMap.get x !entries with
         | Some entry when (not entry.for_type || for_type) -> block
         | _ -> loop blocks)
   in
   loop (!env)
 
 let get_block x block =
-  SMap.find_unsafe x !block
+  let entries = block.entries in
+  SMap.find_unsafe x !entries
 
 let set_block x entry block =
-  block := !block |> SMap.add x entry
+  let entries = block.entries in
+  entries := !entries |> SMap.add x entry
 
 let exists_block x block =
-  SMap.get x !block
+  let entries = block.entries in
+  SMap.get x !entries
+
+let set_env x entry hoist =
+  let rec loop = function
+    | [] -> set_block x entry global_block
+    | block::blocks ->
+        if (not hoist || (hoist && block.scope = Hoist)) then
+          set_block x entry block
+        else
+          loop blocks in
+  loop (!env)
+
+let exists_env x hoist =
+  let rec loop = function
+    | [] -> exists_block x global_block
+    | block::blocks ->
+        if (not hoist || (hoist && block.scope = Hoist)) then
+          exists_block x block
+        else
+          loop blocks in
+  loop (!env)
 
 let update_block ?(for_type=false) x (specific_t, general_t) block =
-  let values = !block in
-  let new_entry = match SMap.get x values with
+  let entries = block.entries in
```

Comment:
Love how this reads now -- so much more clear! ("entries" in a "scope" :ok_hand:)